### ISSUE TITLE
Issue with -p or --threads

### ISSUE THREAD

**Issue by Darkyere** [ckolivas/lrzip#247]
No matter what i do or try i cant make it only use a certain amount of threads.

I am trying to make a test example for a Enhancement propotion for Proxmox sepecifically the BackUp part.

Now my command is as following

`lrzip -L 9 -p 4 -o /Proxmox-BackUp/Manual-vzdump/102-Kubuntu-Plasma-BTRFS-RAID0-lrzip-L9-P4.lrzip`

Now it creates between 8 and 13 threads. 
I allso tried out.

`lrzip -L 7 --threads 4 -o /Proxmox-BackUp/Manual-vzdump/102-Kubuntu-Plasma-BTRFS-RAID0-lrzip-L9-P4.lrzip`

Same result. 8 to 13 Threads.

I am trying to not make it hog to much CPU power by setting a thread count, since proxmox rely on ZFS and is a VM host platform.

It is a dire need to be able to set a Thread count to exactly what a person needs.

Is there any way you can help me with this ?

it is version

```
lrzip --version
lrzip version 0.651
```

Best regards,
Darkyere

**Comment by pete4abw** [comment]
Out of curiosity, try running as root or sudo. May be a permissions thing.

**Comment by Darkyere** [comment]
Funny enough I actually has elavated the terminal with sudo permission because it is needed for the Proxmox vzdump tool which I am piping to lrzip.
So I should already have sudo permissions.

Thank you for looking into it.

**Comment by pete4abw** [comment]
Try not piping the command. Use `lrzip` directly and see.

**Comment by Darkyere** [comment]
I tried not piping just to get an idea if this would make a difference.
Might as well try i suppose, unfortunately it did not work.

![Skærmbillede 2023-08-09 kl  14 05 44](https://github.com/ckolivas/lrzip/assets/23291749/f8090dee-11cf-4749-a60c-da50a5cbd7e2)

I do not know how else to make this example then lrztar.
lrzip requires to pipe through tar and the idea was not to pipe it.

vzdump has no other way than piping the output.

And with both `vzdump... | lrzip...` it seemed to not use the p4 threads as my fist post.

In this example with lrztar
`lrztar -L 9 -p2 -o /home/proxmox/Compress\ example/Test1-lrzip-L9-p4.lrzip /home/proxmox/Compress\ example/Syncthing/`
As the image shows it did not use only 2 threads and 200% CPU usage.
It takes anything that is available to it.

This is done on two seperate debian 12 systems.
To make the test more interesting.
That is why it is with 4 and 2 threads.

Is there anything here that makes sence to you?

**Comment by Darkyere** [comment]
I must say i am embarressed.
I dont sem to be capable to figure out how to pipe through tar.

Does this seem to make sence to you

vzdump 100 --compress 0 --stdout | tar -O -cf - | lrzip -L 9 -p4 -o /Proxmox-BackUp/Manual-vzdump/vzdump-tar-lrzip-L9-p4.lrzip

**Comment by pete4abw** [comment]
Ah, I just remembered, looking at the lzma code. Each backend in `lrzip` may use threads as well. lzma may use up to 32 threads. However, the `-p` option in `lrzip` will control how large the block of data sent to the backend will be. Example: 

**If a chunk of data is 90**:

* And the number of threads is 9, then each data block sent to the backend will be 10 (90/9).
* And the number of threads is 4, then each data block sent to the backend will be 22.5 (90/4).
* And the number of threads is 2, then each data block sent to the backend will be 45 (90/2).

This, of course, if memory and sliding window permits. lrzip or lrzip-next cannot control the backend use of threads over and above what lrzip uses.

The larger the data block, the better backend compression may be because it has more data to look ahead at.

**Comment by Darkyere** [comment]
I know I started by making a mistake by misunderstanding how to pipe it the right way.
But I then edited the original thread yesterday after my mistake.

I did as you just wrote made an vzdump to a .tar file.
And the used lrzip to compress the tar file.

And it actually worked.
It respected the 4 threads as you made a theory about.

Plz. Have a look at the edited post above your last post.

I didn't write the command I used when I edited, maybe I should have.

But your idea worked.
I appreciate your time.
I just do not believe I could ever convince the proxmox team to implement it in this way.

But I still find lrzip as an excellent and useful compression tool.

Bedt regards,
Darkyere

**Comment by pete4abw** [comment]
Yes. it appears that threads are uncontrolled when passed to a backend. This will be an interesting problem since we do not want to constrain the backend from doing its work. I am not sure if this is controllable atm. Good luck, I am done here.

![threads](https://github.com/ckolivas/lrzip/assets/10359685/42626390-42fd-487e-b9cf-3f6254dbb910)
